dot/sync/fullsync.go

```diff
@@ -395,85 +407,79 @@ type RequestResponseData struct {
 	responseData []*types.BlockData
 }
 
-func validateResults(results []*SyncTaskResult, badBlocks []string) (repChanges []Change,
-	peersToBlock []peer.ID, validRes []RequestResponseData) {
-
-	repChanges = make([]Change, 0)
-	peersToBlock = make([]peer.ID, 0)
-	validRes = make([]RequestResponseData, 0, len(results))
-
-resultLoop:
-	for _, result := range results {
-		request := result.request.(*messages.BlockRequestMessage)
+func validateResult(result TaskResult, badBlocks []string) (repChange *Change,
+	blockPeer bool, validRes *RequestResponseData) {
 
-		if !result.completed {
-			continue
-		}
-
-		response := result.response.(*messages.BlockResponseMessage)
-		if request.Direction == messages.Descending {
-			// reverse blocks before pre-validating and placing in ready queue
-			slices.Reverse(response.BlockData)
-		}
+	if !result.Completed {
+		return
+	}
 
-		err := validateResponseFields(request, response.BlockData)
-		if err != nil {
-			logger.Warnf("validating fields: %s", err)
-			// TODO: check the reputation change for nil body in response
-			// and nil justification in response
-			if errors.Is(err, errNilHeaderInResponse) {
-				repChanges = append(repChanges, Change{
-					who: result.who,
-					rep: peerset.ReputationChange{
-						Value:  peerset.IncompleteHeaderValue,
-						Reason: peerset.IncompleteHeaderReason,
-					},
-				})
-			}
+	task, ok := result.Task.(*syncTask)
```

Comment:
What if we define `result.Task` over a generic? so we can skip this casting?

Reply:
I considered making this generic and decided against it, but I can't remember why. Will give it a try.

Reply:
Just tried and the reason I didn't do this is because `SyncService` is supposed to maintain the worker pool and work with any kind of `Strategy`. This won't work if the worker pool is a generic type instantiated with `*syncTask`.

Reply:
Good point, that makes me think that the strategies should be responsible for creating their own workerpool maybe

Reply:
I'm still in favor of having the service manage one pool that is used for all strategies. I think it's easier to manage in terms of concurrency and good separation of concern. Is it worth giving that up to avoid a few casts? The strategies definitely know what concrete type they use as sync tasks.

Reply:
100% agree that the service should be responsible for the working pool, but I'm still trying to find a way to improve the types. We can keep the casts in the meanwhile